sample/system.py

```diff
@@ -1,60 +1,120 @@
+    user_info = {}
+    for i in range(0, len(user_ids), batch_size):
+        user_id_string = ','.join([str(u) for u in user_ids[i:i + batch_size]])
+
+        req = request.Request('%suserinfo?user_id=%s' % (api_url, user_id_string),
+                              headers={"api_key": api_key})
         resp = request.urlopen(req)
-    except error.HTTPError as e:
-        print(e.read())
-    return json.loads(resp.read())
+        user_info.update(json.loads(resp.read())['userinfo'])
+    return user_info
 
 
-def sendRecommendations(data, api_key):
-    data = json.dumps({'recommendations': data}).encode('utf8')
-    req = request.Request(link+"recommendation", data=data, headers={
-                          'Content-Type': 'application/json', "api_key": api_key})
-    resp = request.urlopen(req)
+def get_articles_by_keywords(keywords, index, window_size=1, size=10000):
+    """Retrieves articles from the Elasticsearch index mentioning 'keywords',
+    the 'window_size' is the number of days back in time articles will be included from."""
+    es = Elasticsearch()
+    query = {
+        'query': {
+            'bool': {
+                'must': {
+                    'match': {
+                        'catch_all': {
+                            'query': keywords,
+                        }
+                    }
+                },
+                'filter': {
+                    'range': {
+                        'date': {
+                            'gte': 'now-{}d'.format(window_size)
+                        }
+                    }
+                }
+            }
+        }
+    }
+    res = es.search(index=index, body=query, params={"size": size})
+    return res
+
+
+def send_recommendations(recommendations, api_key, api_url, batch_size=100):
+    """Sends the recommendations to the arXivDigest API. Recommendations should be a dict of user_id, recommendation pairs."""
+    recommendations = list(recommendations.items())
+    for i in range(0, len(recommendations), batch_size):
+        data = json.dumps({'recommendations': dict(recommendations[i:i + batch_size])}).encode('utf8')
+        req = request.Request(api_url + "recommendation", data=data, headers={
+            'Content-Type': 'application/json', "api_key": api_key})
+        resp = request.urlopen(req)
+        if resp.getcode() is not 200:
+            print(resp)
+
+
+def make_recommendations(user_info, index, n_articles=10):
+    """Makes recommendations for all the users in user_info based on the keywords in user_info.
+    Searches the elasticsearch index for candidates and uses the Elasticsearch score as score.
+    'n_articles' is the number of articles to recommend for each user."""
+    recommendations = {}
+    for user, info in user_info.items():
+        articles = get_articles_by_keywords(' '.join(info['keywords']), index, window_size=1)['hits']['hits']
+        article_recommendations = []
```

Comment:
This means that the same article may be recommended multiple times.
Instead, have a dict indexed with article_id where you accumulate the scores (an article may match multiple keywords, in which case it should have a higher score).

Also, the explanation should be like "This article matches the keywords 'X' [Y, Z] that you are interested in."

Reply:
Might have misunderstood how you meant the simple system should work when we talked, but now I've changed it such that it should be more similar to what you described here. Would you check that its more like you meant know.

The keyword matching is still quite simple though meaning that for example the keyword "entity oriented search" will also match all documents containing "entity".
We could require all terms to match by changing the operator to AND instead of OR or require a percentage of the terms to match by using minimum_should_match. Is this something we would want?

Reply:
Using (default) OR matching is fine.  Documents matching more query terms will receive a higher score.